dev-docs/RFCs/v5.0/attribute-animation-rfc.md

```diff
@@ -9,41 +9,128 @@ Notes:
+
+If interpolation is done on the CPU: both calculating the attribute and uploading it to the GPU will be expensive. Framerate is expected to stagger.
+
+If using TransformFeedback:
+- At initialization: there is a cost to compile shaders and create the model for the transform feedback. We should minimize it by only doing it for the layers that have transition enabled.
+- At attribute update: new value arrays are uploaded to GPU once at the beginning of the animation, which is the same as today. There is a cost to calculating and uploading the "current" state of the attribute to animate from.
+- During animation: The transform feedback is updated every frame till the end of the transition. It renders to buffers that are directly transferred to the vertex shader, so perf cost should be minimal.
+
+
+## Questions
+
+### Enter and exit behaviors
+
+Enter (appearance of new geometries) may happen when:
+**A** The layer is added or becomes visible
+**B** The data array is changed with increased size
+
+Exit (disappearance of rendered geometries) may happen when:
+**C** The layer is removed or becomes invisible
+**D** The data array is changed with decreased size
+
+Without adding additional complexity, the most straightforward behavior is no enter/exit animation - geometries appear and disappear immediately. The application can create their own enter/exit animations by adding invisible objects (e.g. instead of removing objects from the data array, change their colors to transparent).
+
+To support enter/exit animations, we may consider one of the following:
+- Add a field `voidValue` to attribute definitions. This could be a value or an transform to be used for enter animation to transition from, or exit animation to transition to.
 ```js
-  new Layer({color: [255, 0, 0]}) - would set the uniform once, as today
+this.state.attributeManager.add({
+  // Animatable
+  radius: {size: 1, accessor: 'getRadius', update: this.calculateRadius, animate: true, voidValue: 0},
+  colors: {size: 4, type: GL.UNSIGNED_BYTE, accessor: 'getColor', update: this.calculateColors, animate: true, voidValue: ([r, g, b, a]) => [r, g, b, 0]}
+});
 ```
-while
+- Add fields `enter` and `exit` to animation parameters. The attribute updater is called with these functions substituting the regular accessor to retrieve the from value for enter animation, or to value for exit animation. This allows the user to control the animations per object, but more expensive at run time:
 ```js
-  new Layer({color: ({tick}) => [tick % 255, 0, 0]}) - would activate animation of this layer and update this prop every frame with an incremented tick value.
+new Layer({
+  transition: {
+    getColors: {
+      duration: 300,
+      enter: feature => feature.properties.fill.concat(0),
+      exit: feature => feature.properties.fill.concat(0)
+    }
+  }
+});
 ```
 
+To properly support scenario **B** and **D**, we need a way to diff data arrays to determine which objects are added/removed.
```

Comment:
Fine for now, although I'd like to discuss these scenarios a bit more further on.  There are complications when updating non-instanced layers like a path layer, a single path can change length messing up the interpolation of all the following verts if special care is not taken. These problems relate to much of the thinking behind my RFC on partial updates.

Reply:
Admittedly it would be nice to have these, though we are circumventing these scenarios by calling the feature "transition" (i.e. instant enter and exit)